### PR TITLE
Add type meta information generation

### PR DESCRIPTION
**Pull request by reicheratwork**
This allows CycloneDDS to do assignability checks during the discovery process, in stead of just looking at the type name.

This PR does require [PR1147](https://github.com/eclipse-cyclonedds/cyclonedds/pull/1147) on CycloneDDS to be merged, as MSVC-C++ does not like certain C features being used (zero-size arrays), which is now being pulled in to the compiled code.

### CODE REVIEW

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/TopicTraits.hpp [hunk 1]
```diff
@@ -18,72 +18,49 @@
...
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/mh3.h"
+#include "dds/ddsrt/md5.h"
+#include "dds/ddsi/q_radmin.h"
+#include "dds/ddsi/ddsi_serdata.h"
+#include "dds/ddsi/ddsi_sertype.h"
+#include "dds/ddsi/ddsi_typelib.h"
+#include "dds/ddsi/ddsi_cdrstream.h"
+#include "org/eclipse/cyclonedds/core/cdr/cdr_enums.hpp"
+#include "dds/features.hpp"
```
[the remark this answers is not in the document]
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/TopicTraits.hpp:31]: This was mainly due to me copying the includes from where the type info functions were implemented in cyclonedds.
I am currently looking in to which includes can be trimmed from this.

**Comment by reicheratwork** [comment]
Managed to trim quite a few includes after looking closely.

**Comment by eboasson** [comment]
@reicheratwork there are three things I would like to ask you to take a look at:

The first is that there is now a conflict with `master`, I think it was my merging a hashing fix that causes the conflict but a quick look suggests you shuffled stuff around so I'd rather you took a look at it then mess it up myself.

Second is that Clang for me warns on a doxygen comment: https://github.com/eboasson/cyclonedds-cxx/commit/1d6a083dc2c08a62c6c65a0e33efb496692b8a11 addresses it, I think it'd make sense to include it.

Finally I didn't realise the `q_radmin.h` include had been in `datatopic.hpp` for a while now (why I did not realise that is a mystery, but whatever), but I really dislike that one getting pulled into application code, not in the least because it pulls in lots of junk into the name space. Considering that its presence in `datatopic.hpp` is not new, I am not going to insist on changing that now.

Still, not having realised that it was there already, I gave moving it out a shot: https://github.com/eboasson/cyclonedds-cxx/commit/a24f49338f1277b4c716ecdad13792edb15539a3. It isn't the commit that makes me the most proud, but if you could look at it to see whether it is good enough to be merged in (or altered in some form because of name spaces, file names, locations, ...) that'd be great.

**Comment by reicheratwork** [comment]
Going over your comments one-by-one:
- the conflict is caused by the large restructuring by me of the different ops functions and your small change to one of them, this should be easy enough to fix
- the wrong parameter name was probably caused by me implementing the doxygen code in a previous iteration of this code, the C interface changing the names of parameters etc, and me changing the functions with it, but leaving the documentation the same, easy fix
- the q_radmin.h include is indeed an ugly one, but this one is necessary due to that file defining things like the fragchain struct (nnrdata) and others, I will take a look at your suggestion (a24f493)

**Comment by reicheratwork** [comment]
seems like that if we want to not expose all bare C code in q-radmin.h by just including everything without a care, some kind of wrapper like your suggestion is indeed necessary
I have therefore cherry-picked your commit as the solution to this issue

**Review by eboasson — APPROVED** [review]
Thanks @reicheratwork !

**Pull request merged**